sstables_loader.cc

```diff
@@ -488,20 +491,69 @@ class sstables_loader::download_task_impl : public tasks::task_manager::task::im
+    // we sync the background abort call.
+    impl::abort();
+    _aborting = smp::invoke_on_all([this] {
+        // Note: subscriptions on every abort source will be local to that shard (intrusive collection)
+        // These are removed by scope end in each processing below.
+        // This call will set an exception pointer. Use pointer from our source. This belongs on this shard,
+        // but the other shards should not retain it across our retention -> we free it.
+        _shard_aborts[this_shard_id()].request_abort_ex(_as.abort_requested_exception_ptr());
+    });
+}
+
 future<> sstables_loader::download_task_impl::run() {
     // Load-and-stream reads the entire content from SSTables, therefore it can afford to discard the bloom filter
     // that might otherwise consume a significant amount of memory.
     sstables::sstable_open_config cfg {
         .load_bloom_filter = false,
     };
     llog.debug("Loading sstables from {}({}/{})", _endpoint, _bucket, _prefix);
-    auto [ table_id, sstables_on_shards ] = co_await replica::distributed_loader::get_sstables_from_object_store(_loader.local()._db, _ks, _cf, _sstables, _endpoint, _bucket, _prefix, cfg);
-    llog.debug("Streaming sstables from {}({}/{})", _endpoint, _bucket, _prefix);
-    co_await _loader.invoke_on_all([this, &sstables_on_shards, table_id] (sstables_loader& loader) mutable -> future<> {
-        co_await loader.load_and_stream(_ks, _cf, table_id, std::move(sstables_on_shards[this_shard_id()]), false, false);
+    auto [ table_id, sstables_on_shards ] = co_await replica::distributed_loader::get_sstables_from_object_store(_loader.local()._db, _ks, _cf, _sstables, _endpoint, _bucket, _prefix, cfg, [this] {
+        return &_shard_aborts[this_shard_id()];
     });
+    llog.debug("Streaming sstables from {}({}/{})", _endpoint, _bucket, _prefix);
+    std::exception_ptr ex;
+    try {
+        _as.check();
+        co_await _loader.invoke_on_all([this, &sstables_on_shards, table_id] (sstables_loader& loader) mutable -> future<> {
+            co_await loader.load_and_stream(_ks, _cf, table_id, std::move(sstables_on_shards[this_shard_id()]), false, false);
+        });
+    } catch (...) {
+        ex = std::current_exception();
+    }
+
+    if (_as.abort_requested()) {
+        // TODO: cleanup. We should not be leaving partial files or anything (?), since mutation
+        // sources will be s3 remotes. But we might now have a partial table state. Is this the
+        // expected result?
+        auto f = std::exchange(_aborting, make_ready_future<>());
+        // wait for any shard-abort to complete
+        co_await std::move(f);
+        if (!ex) {
+            ex = _as.abort_requested_exception_ptr();
+        }
+    }
+
+    if (ex) {
+        co_await _loader.invoke_on_all([&sstables_on_shards] (sstables_loader&) {
+            sstables_on_shards[this_shard_id()] = {}; // clear on correct shard
```

Comment:
Wasn't it std::move()-d up above by
```
co_await loader.load_and_stream(_ks, _cf, table_id, std::move(sstables_on_shards[this_shard_id()]), false, false);
```
?

Reply:
Yes. If we get that far. But it is possible we abort before even reaching the previous invoke_on_all.